src/liballoc/string.rs

```diff
@@ -1419,6 +1425,7 @@ impl String {
     /// assert_eq!(world, "World!");
     /// # }
     /// ```
+    #[must_use]
```

Comment:
This function *does* have side effects (`&mut self`) — why should the user need to care about the return value here? What should they use instead of ignoring the result?

Reply:
If I ignore the result of `split_off`, it seems that I should better use `truncate`, shouldn't I?

Reply:
Is there a way to customize the error message provided by `#[must_use]` on a case-by-case basis to suggest that `truncate` is more appropriate? Otherwise, should that go into the documentation?

Reply:
The attribute can take a string literal with an additional message: [usage](https://github.com/rust-lang/rust/blob/fb9e0588901dc2311f5e8f63f1c041e83041a596/src/test/ui/fn_must_use.rs#L28), [output](https://github.com/rust-lang/rust/blob/master/src/test/ui/fn_must_use.stderr#L19). (The output format may [change soon](https://github.com/rust-lang/rust/pull/50437).)

Reply:
This one I like provided it has the message.  It sets a nice "if it allows redirecting to a cheaper method" precedent (which, in retrospect, [`collect`](https://github.com/rust-lang/rust/pull/49533/files#diff-1398f322bc563592215b583e9b0ba936R1371) is also following).

(Should probably be on `Vec::split_off` too, though.)